Expose security levels

mhdawson opened the pull request:
<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

mhdawson commented:
Run which tests on the different OpenSSL versions - https://ci.nodejs.org/job/node-test-commit-linux-containered/48401/

All passed, so I think that confirms it is working correctly.

In terms of testing I don't think we can change the default except at compile time, and I also  think comparing against a specific version could cause problems for the shared library testing.

I can add a test that just calls the method and makes sure it is within the range documented which is currently 1-5. Does that makes sense to people?

jasnell commented on doc/api/crypto.md:
```diff
@@ -4165,6 +4165,16 @@ const {
+
+* Returns: {number} The [`default OpenSSL security level`][].
+  Useful to be able to adjust based on the algorithms available given the
+  security level.
```
Some explanation as to how to interpret the returned values would be helpful. 

And just for my knowledge because I'm not that familiar with the security levels... are the return values specific at all the version of OpenSSL? Is it supported by BoringSSL?

jasnell replied:
Ok just digging into the API a bit more ... https://docs.openssl.org/3.0/man3/SSL_CTX_set_security_level/#description 

Looks like there are six security levels 0 through 5.... honestly we should document these and/or link to the relevant openssl definitions of them here.

richardlau replied:
> And just for my knowledge because I'm not that familiar with the security levels... are the return values specific at all the version of OpenSSL? Is it supported by BoringSSL?

Yes, the values mean different things for different OpenSSL versions.

Compare the level descriptions for 1.1.1, 3.0 and 3.4:
- https://docs.openssl.org/1.1.1/man3/SSL_CTX_set_security_level/#default-callback-behaviour
- https://docs.openssl.org/3.0/man3/SSL_CTX_set_security_level/#default-callback-behaviour
- https://docs.openssl.org/3.4/man3/SSL_CTX_set_security_level/#default-callback-behaviour

jasnell replied:
This makes me more nervous about this API as users will need to know exactly which version of openssl the version of node.js they are using happens to have been compiled with. This absolutely needs to be documented and explained.

richardlau replied:
Maybe this should be an internal only API that is used for the tests?

There's definite value in being to get this information and use it to adjust test expectations to make it easier for downstream rebuilders of Node.js (e.g. Linux distros) to run the tests when Node.js is linked to a system OpenSSL. I'm less convinced of the usefulness for general Node.js applications.

jasnell replied:
I'd be happy with that. I really can't think of much of a general use case for this outside of our tests.

jasnell commented on src/crypto/crypto_util.cc:
```diff
@@ -196,6 +196,23 @@ void TestFipsCrypto(const v8::FunctionCallbackInfo<v8::Value>& args) {
+    return ThrowCryptoError(env, ERR_get_error(), "SSL_new");
+  }
+
+  int sec_level = SSL_get_security_level(ssl);
```
consider adding a method to `ncrypto::SSLPointer`?

jasnell commented on lib/crypto.js:
```diff
@@ -44,6 +44,7 @@ const { getOptionValue } = require('internal/options');
 const {
   getFipsCrypto,
   setFipsCrypto,
+  getOpenSSLSecLevelCrypto,
```
```suggestion
  getOpenSSLSecLevelCrypto: getOpenSSLSecLevel,
```

jasnell commented on lib/crypto.js:
```diff
@@ -256,6 +258,10 @@ function setFips(val) {
   }
 }
 
+function getOpenSSLSecLevel() {
+  return getOpenSSLSecLevelCrypto();
+}
+
```
Can remove the indirection through the function by aliasing the import 

```suggestion
```

jasnell commented on test/parallel/test-tls-dhe.js:
```diff
@@ -43,7 +44,7 @@ const dheCipher = 'DHE-RSA-AES128-SHA256';
 const ciphers = `${dheCipher}:${ecdheCipher}`;
 
-if (!common.hasOpenSSL(3, 2)) {
+if (secLevel < 2) {
```
I know this was a problem with the original code here also but it would be helpful to add a comment explaining what the `2` here is.... a pointer to the docs would suffice.

mhdawson commented:
> Yes, the values mean different things for different OpenSSL versions.

@richardlau what are the differences in meaning between the openssl levels ?

They all seem to say 80, 112, 128, 192 and 256 bits in terms of what the security levels map to.  The default level might be different but that is what the API exposes.

mhdawson commented:
Added the test since I had it ready to go.

mhdawson commented:
In terms of doc  Returns: {number} The [`default OpenSSL security level`][]. is a link to the description of the security levels.

mhdawson commented:
The main question though is if it should just be private method, if that seems to be the consensus are there suggestions of where I should best add it?

richardlau commented:
> > Yes, the values mean different things for different OpenSSL versions.
> 
> @richardlau what are the differences in meaning between the openssl levels ?
> 

For example. compare the [Level 1 description for OpenSSL 3.0]:(https://docs.openssl.org/3.0/man3/SSL_CTX_set_security_level/#default-callback-behaviour)
> The security level corresponds to a minimum of 80 bits of security. Any parameters offering below 80 bits of security are excluded. As a result RSA, DSA and DH keys shorter than 1024 bits and ECC keys shorter than 160 bits are prohibited. All export cipher suites are prohibited since they all offer less than 80 bits of security. SSL version 2 is prohibited. Any cipher suite using MD5 for the MAC is also prohibited. Note that signatures using SHA1 and MD5 are also forbidden at this level as they have less than 80 security bits.

to [OpenSSL 3.4](https://docs.openssl.org/3.4/man3/SSL_CTX_set_security_level/#default-callback-behaviour):
> The security level corresponds to a minimum of 80 bits of security. Any parameters offering below 80 bits of security are excluded. As a result RSA, DSA and DH keys shorter than 1024 bits and ECC keys shorter than 160 bits are prohibited. Any cipher suite using MD5 for the MAC is also prohibited. Any cipher suites using CCM with a 64 bit authentication tag are prohibited. Note that signatures using SHA1 and MD5 are also forbidden at this level as they have less than 80 security bits. Additionally, SSLv3, TLS 1.0, TLS 1.1 and DTLS 1.0 are all disabled at this level.

OpenSSL 3.4 adds disabling  SSLv3, TLS 1.0, TLS 1.1 and DTLS 1.0 at Level 1. In OpenSSL 3.0, SSLv3 is disabled at Level 2, TLS 1.0 at Level 3 and TLS 1.1 at Level 4. Also the OpenSSL 3.4 mentions "Any cipher suites using CCM with a 64 bit authentication tag are prohibited", which is not present in the description for OpenSSL 3.0.

mhdawson commented:
Looks like failures were due to changes landed since I rebased yesterday.

mhdawson commented:
@richardlau thanks for pointing out the specific differences. Seems odd that they would change what a security level means but some of those do seem like functional changes.

mhdawson commented:
@jasnell I think I addressed your comments as well as moving the method so it is internal only. 

In terms of boringSSL one remaining question I had was how version numbers are reported in process.versions when boringssl is used because the existing checks:

```JavaScript
const hasOpenSSL = (major = 0, minor = 0, patch = 0) => {
  if (!hasCrypto) return false;
  if (OPENSSL_VERSION_NUMBER === undefined) {
    const regexp = /(?<m>\d+)\.(?<n>\d+)\.(?<p>\d+)/;
    const { m, n, p } = process.versions.openssl.match(regexp).groups;
    OPENSSL_VERSION_NUMBER = opensslVersionNumber(m, n, p);
  }
  return OPENSSL_VERSION_NUMBER >= opensslVersionNumber(major, minor, patch);
};
```

Only seem to consider openSSL, unless process.versions still look like OpenSSL is being used even though its boringssl under the covers. Do you know how the versions are reported when boringssl is used?

mhdawson commented:
@richardlau are you ok with the PR after the latest updates?

jasnell commented:
@mhdawson .... I'm not sure how boringssl is represented in process.versions. This might be helpful https://github.com/search?q=repo%3Aelectron%2Felectron+%22process.versions%22+%22boringssl%22&type=code

mhdawson commented:
@jasnell, it looks like that means that if BoringSSL is used then process.vesions.openssl is not defined. In that case none of the existing checks being moved over to the check against the security level should help/hurt the BoringSSL case.

jasnell commented:
I would confirm that with the electron team before progressing.  @codebytere

richardlau commented on src/crypto/crypto_util.cc:
```diff
@@ -201,6 +203,27 @@ void TestFipsCrypto(const v8::FunctionCallbackInfo<v8::Value>& args) {
 
+void GetOpenSSLSecLevelCrypto(const FunctionCallbackInfo<Value>& args) {
+  // for BoringSSL assume the same as the default
+  int sec_level = 1;
```
Maybe this should be `OPENSSL_TLS_SECURITY_LEVEL`?
https://github.com/nodejs/node/blob/7bc2946293757389468f1fa09714860f8e1147b7/deps/openssl/openssl/include/openssl/tls1.h#L29-L32

While `1` is the default for OpenSSL 3.0 and 3.1, [OpenSSL 3.2 changed the default to `2`](https://docs.openssl.org/3.2/man3/SSL_CTX_set_security_level/#notes).

mhdawson replied:
```suggestion
  int sec_level = OPENSSL_TLS_SECURITY_LEVEL;
```

mhdawson commented:
@codebytere let us know if you have any concerns.

jasnell commented on test/parallel/test-crypto-sec-level.js:
```diff
@@ -0,0 +1,11 @@
+const assert = require('assert');
+
+const secLevel = require('internal/crypto/util').getOpenSSLSecLevel();
+assert.ok(secLevel >= 0 && secLevel <= 5);
```
A comment explaining these numbers and where they come from would be helpful for future generations coming into this code later.

mhdawson commented on test/parallel/test-tls-dhe.js:
```diff
@@ -43,7 +45,7 @@ const dheCipher = 'DHE-RSA-AES128-SHA256';
 const ciphers = `${dheCipher}:${ecdheCipher}`;
 
-if (!common.hasOpenSSL(3, 2)) {
+if (secLevel < 2) {
```
I don't think we want to explain security levels everywhere but adding a comment like 

"For more information on openSSL security levels see - https://docs.openssl.org/1.1.1/man3/SSL_CTX_set_security_level/#default-callback-behaviour"

every time they are used might be feasible. Would that cover what you had in mind ?

mhdawson replied:
Another idea I had is how about we add that comment to where we get the secLevel. If we do it once when we get the secLevel then we might add a bit more info.

I'll try that out and add to the PR

jasnell replied:
Making the `2` into a constant at the top of the file and adding the comment there would also work.

mhdawson commented:
Test that failed in first run is already marked as flaky, resuming.

mhdawson commented:
Rebased

mhdawson commented:
pushed commit to address linter

mhdawson commented:
Landed in f2d274753a2a
